graph/src/components/ethereum/network.rs

```diff
@@ -0,0 +1,227 @@
+use failure::{format_err, Error};
+use std::cmp::{Ord, Ordering, PartialOrd};
+use std::collections::HashMap;
+use std::fmt;
+use std::sync::Arc;
+
+use crate::components::ethereum::EthereumAdapter;
+pub use crate::impl_slog_value;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct NodeCapabilities {
+    pub archive: bool,
+    pub traces: bool,
+}
+
+impl Ord for NodeCapabilities {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match (
+            self.archive.cmp(&other.archive),
+            self.traces.cmp(&other.traces),
+        ) {
+            (Ordering::Greater, Ordering::Greater) => Ordering::Greater,
+            (Ordering::Greater, Ordering::Equal) => Ordering::Greater,
+            (Ordering::Equal, Ordering::Greater) => Ordering::Greater,
+            (Ordering::Equal, Ordering::Equal) => Ordering::Equal,
+            (Ordering::Less, _) => Ordering::Less,
+            (_, Ordering::Less) => Ordering::Less,
+        }
+    }
+}
+
+impl PartialOrd for NodeCapabilities {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl fmt::Display for NodeCapabilities {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            NodeCapabilities {
+                archive: true,
+                traces: true,
+            } => write!(f, "archive, trace"),
+            NodeCapabilities {
+                archive: false,
+                traces: true,
+            } => write!(f, "full, trace"),
+            NodeCapabilities {
+                archive: false,
+                traces: false,
+            } => write!(f, "full"),
+            NodeCapabilities {
+                archive: true,
+                traces: false,
+            } => write!(f, "archive"),
+        }
+    }
+}
+
+impl_slog_value!(NodeCapabilities, "{:?}");
+
+#[derive(Clone)]
+pub struct EthereumNetworkAdapter {
+    pub capabilities: NodeCapabilities,
+    adapter: Arc<dyn EthereumAdapter>,
+}
+
+#[derive(Clone)]
+pub struct EthereumNetworkAdapters {
+    pub adapters: Vec<EthereumNetworkAdapter>,
+}
+
+impl EthereumNetworkAdapters {
+    pub fn cheapest_with(
+        &self,
+        required_capabilities: &NodeCapabilities,
+    ) -> Result<&Arc<dyn EthereumAdapter>, Error> {
+        let sufficient_adapters: Vec<&EthereumNetworkAdapter> = self
+            .adapters
+            .iter()
+            .filter(|adapter| &adapter.capabilities >= required_capabilities)
+            .collect();
+        if sufficient_adapters.is_empty() {
+            return Err(format_err!(
+                "A matching Ethereum network with {:?} was not found.",
+                required_capabilities
+            ));
+        }
+        // TODO: Randomly choose between capable nodes
```

Comment:
Looks like this should make it easy: https://docs.rs/rand/0.7.3/rand/seq/trait.IteratorRandom.html

Reply:
@fordN I think the comment can go